### ISSUE TITLE
Optimisation: normalising `bled_codes` only once

### ISSUE THREAD

**Issue by kdharris101** [reillytilbury/coppafish#81]
Seems like we are normalizing the bled codes in the inner loop of omp.  We might speed things up by doing this just once at the beginning.

https://github.com/reillytilbury/coppafish/blob/3278eede3410ea7066752ec7d2407c826f8bd165/coppafish/call_spots/dot_product_optimised.py#L32

**Comment by paulshuker** [comment]
I agree this would be faster, we just need to check the change will not have unexpected consequences, like maybe the bled codes are left unnormalised outside the dot product function for a reason? I am unsure.

**Comment by kdharris101** [comment]
Can't see what that reason would be: I think the algorithm is invariant to scaling of each gene's bled code; and with the new scaling we will L2 normalize each gene's free code (which takes the place of the bled code) after computing it. But agree we need to sanity-check that it produces the same results as the old version. 

Of all the things that might speed up OMP i think this is the most promising. Dot_product_score() is run once for every spot, and each time it normalizes the bled codes of every gene, which is about twice as much computation as computing the dot product of all genes with the spot colors.  I would guess that dot_product_score() takes the most running time of OMP, so this change could potentially reduce running time of OMP to 1/3 previous.  (Unless jax is doing something clever by not reruning a computation it has already done on an unchanged array.)

**Comment by paulshuker** [comment]
You might be right that the dot product score takes the most time, it is hard to track with jax on so many nested functions. But I think the calls are made inside the pmap if I remember correctly. This might be a significant speed up, although the actual computation of OMP is incredibly fast now, the slow part is actually loading in the spot colours and doing computations on loaded data before reaching the jax pmap.

**Comment by kdharris101** [comment]
Ah OK.  Within OMP yes I would guess dot product takes longest since it is order n_pixels x n_genes on each iteration.  The other main operation is finding coefficients which is order n_pixels x n_genes_used, which is substantially less.  Yes it is inside pmap but still is the heaviest computation for OMP.

But if all of OMP isn't taking much time, this is irrelevant.  

When you say loading spot colors do you mean disk access is the bottleneck?  

And what are the computations on loaded data you are referring to? Do you know which function?

**Comment by paulshuker** [comment]
So when I say loading spot colours I mean the progress bar that says `loading` X number of spot_colours...` takes about half the time or more. I struggle to believe that the disk reading is actually that slow, so I think it does some other computations before reaching the jax code, like an intensity thresholding perhaps, or maybe some axis switching in numpy? I need to look into this code more later on and locate the slow lines of code with a profiler.
